### ISSUE TITLE
block: Implement rough sync logic

### ISSUE THREAD

**Issue by renaynay** [celestiaorg/celestia-node#78]
A full node should be able to sync against the `Fetcher` (whether the fetcher is a core node or another celestia full node).

A full node, once started, should do the following:

* `GetBlock` with nil set as height (in order to get latest block @ latest height) 
* Sync backwards using `GetBlock(height)` until the entire chain is synced up to last block that node knows of

**Question**: for this rough sync version, should we allow `BlockService` to listen for new blocks simultaneously while syncing?

**Comment by adlerjohn** [comment]
Process seems fine to me as a start. Quick Q: would the weak subjectivity checkpoint be handled by the Celestia Node or Core Node?

>  for this rough sync version, should we allow `BlockService` to listen for new blocks simultaneously while syncing?

I don't think that's strictly necessary, but also wouldn't hurt. Without that feature you might have to do sync a few times until you're sync'ed to the tip though?

**Comment by liamsi** [comment]
> can you pls point me to the code in Tendermint which implements syncing from the week subjectivity checkpoint?

Tendermint is only initialized subjectively as a light client or when state syncing. There are no checkpoints for regular full nodes (not sure if that is what @adlerjohn meant though).

Here are the places:
- https://github.com/tendermint/tendermint/blob/66a11fe52795eaba6aacae3ca25206c380e5283a/light/client.go#L167 and https://github.com/tendermint/tendermint/blob/66a11fe52795eaba6aacae3ca25206c380e5283a/light/trust_options.go#L32-L35 for the light client
- similarly state sync gets initialized subjectively with a trusted hash and height: https://github.com/tendermint/tendermint/blob/ded310093e0d771c9ed27f296921cb6b23d99f29/internal/statesync/reactor.go#L185-L186 https://github.com/tendermint/tendermint/blob/86f00135dd5d14a382a033c705062839d75e66cf/config/config.go#L804-L806

**Comment by Wondertan** [comment]
I see, so we should do the initial subjective sync using `statesync` for light-node and full blockchain sync with `blocksync` for full node. Thanks @liamsi 

Alhought, does the light-node should have the whole celestia network state ?

cc @adlerjohn

**Comment by liamsi** [comment]
> @liamsi is there a way to start a discussion with the Tendermint devs on upstreaming a change to that effect?

Yes, absolutely. Just talked to @cmwaters about it and he might have some interesting ideas (which would involve some changes in the SdK though). I don't think it's of high priority right now though tbh. 

Either way, the change can definitely be upstreamed. You can formulate your suggestion in an issue first.

**Comment by Wondertan** [comment]
Synchronization is done through Headers
